### DISCUSSION TITLE
Jax pytrees with tuples/lists/etc as leaf nodes

### DISCUSSION THREAD

**Discussion by jbuckman** [jax-ml/jax#12826] [Q&A]
Is there a way to use jax pytrees with tuples/lists/etc on the leaves? Concretely, I want to do a function like this:
```
def append_gradient(gradient_accumulator, gradient):
    return jax.tree_map(lambda ga, g: ga + (g,), gradient_accumulator, gradient)
```
where `gradient_accumulator` is a pytree containing python tuples at each leaf, and `gradient` is a pytree with the same structure, containing a DeviceArray at each leaf. This `tree_map` call fails, because it confuses the tuple on the `gradient_accumulator` (which I want to be the leaf) for part of the pytree structure, and throws an error because `gradient` does not have a tuple in that same place.

I've also tried working around this by wrapping my tuple in a custom class, but this failed with
```TypeError: Argument '<__main__.LeafTuple object at 0x7f89f4475730>' of type <class '__main__.LeafTuple'> is not a valid JAX type.```

Is what I am attempting to do even possible? Or is there some restriction against having variable-sized leaves on a pytree? In my case, I can't just pre-allocate a fixed-size DeviceArray because in my algorithm, I don't know in advance how many gradients I'll need to accumulate. Although, if needed I can workaround this issue by just pre-allocating and including expanding the size if needed.

**Comment by sholtodouglas** [comment]
There are two things you can do here:

 - Make a custom class which inherits from tuple, and then use that in place of the tuple. The map will not traverse into the custom class, instead treating it all as one leaf
 - Put gradient as the first argument, because tree_map uses the first argument to inform the tree structure

Hope that helps!

**Reply by froystig** [reply]
I typically go with @sholtodouglas' second recommendation in this situation. Just to give a complete answer to this part too:

> I've also tried working around this by wrapping my tuple in a custom class, but this failed [...]

If this is a new class you've defined (e.g. not inheriting from `tuple` as in @sholtodouglas' first recommendation), you would need to register your custom class as a pytree type (via [1](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.register_pytree_node.html) or [2](https://jax.readthedocs.io/en/latest/_autosummary/jax.tree_util.register_pytree_node_class.html)).
